### DISCUSSION TITLE
Line squaring with blocks

### DISCUSSION THREAD

**Discussion by Schillil2001** [pybricks/support#1699] [General]
I'm helping with my daughter's FLL team this fall. They are 7th graders who have not learned any Python yet. I'll admit, I'm not familiar with Python either which is why I bought 3 copies of the Block version. So far it has been wonderful. That said, anyone who has seen the trailer for the FLL: Submerged board for this season will note that line following won't be too helpful, but line squaring is going to be crucial. I messed with it a little last evening using a Technic bot that I put together just to practice the basics and it isn't working. I used dark blue poster board, then marked it with white then black so that it resembles the way this year's board will look. For reference, I can't get it to work well with the regular Spike software as well. I'm a pharmacist by trade and very novice with computer programming so any help would be appreciated.

**Comment by laurensvalk** [comment]
Thank you for using block coding and opening a discussion here! 

Can you describe in words or bullet points the process of 'line squaring', for those of us less familiar with FLL? How does the robot start, and how should it end up? How many sensors does the robot have, and how are they positioned? Is the robot allowed to cross the line temporarily during the operation, or is there no space for that? 

The more detail, the better. A picture of last year's board or the poster board you made might also help. Or even a video where you move the robot by hand. This way it might become a bit of a recipe for what you can then implement in blocks.

Not trying to be pedantic - this tends to really help, including for your students and others finding this discussion thread later :)

**Reply by Schillil2001** [reply]
Ok, I'll try. In FLL, they used to use line following, but they have moved to more line squaring. In front of a mission there will be a pair of lines. First a white one then black. The white is there because the boards are becoming more and more colorful.  So, using a pair of color sensors, the hub needs to "see" white then black. When one of them encounters the black lines after the white the robot needs to stop and adjust automatically based on when the second sensor sees it. This permits the robot to line up. In the old Spike 2 legacy software you had it move back and forth a couple times to make sure it was directly perpendicular to the mission model. It is easier, then, to move directly forward and interact with the model.

Not sure if it will work, but I'm attempting to attach a close up picture of this year's board with the the parts of
![submerged board close-up](https://github.com/pybricks/support/assets/154741252/7c3a50c5-a4ae-40a2-9277-f4c0a798ed80)

**Reply by Schillil2001** [reply]
The biggest issue we had last year was that the color variability of the boards. Last year the blue, purple and black all seemed to "read" as black and our robot would start acting strangely if it misunderstood the input from the sensors. And me? I had no clue. I'm just a parent volunteer who took Intro to C++ 25 years ago on a whim when I had a spare hour that needed to be filled to maintain fulltime status in college. That was 1997, so I'm pretty rusty. And I don't do any programming anymore since I'm a pharmacist.

**Reply by laurensvalk** [reply]
Thank you, that is helpful. How wide are those black and white lines in front of the missions typically? Making a half turn as in my video probably wouldn't quite fit. 

I'm not up to speed on the FLL calandar. If there's time, you could perhaps still have your team experiment with more standalone examples like the one posted below. Even if it isn't a direct match, it can be good practice.

As far as the color sensor goes, you can see a suggestion for calibrating it in the other thread as well.

**Reply by laurensvalk** [reply]
Likewise, if there's time, it can sometimes also be a good activity to have your team brainstorm about the plan for the robot before getting into the code. If you encourage your students to record their thought process, this usually also works well with the judges.

**Reply by Schillil2001** [reply]
They don't even release the board for shipping until Aug 6 when they do the full reveal. This is from a screen capture of the teaser trailer. I want to say, if it is like last year, they are between half and 3/4 of an inch thick.

The team doesn't get together before school starts in September. My 12 year old and I are just trying to work out the odd programming kinks now while we have the time. And I'm not the main coach. Technically, I'm not even a coach, just an extra parental pair of hands to help keep seven 12 year old kids (4 girls and 3 boys) heading in the same general direction. This is a new team for my daughter because her old team from last year was pretty dysfunctional. They had a habit of either ignoring her ideas completely or discounting them since she was the only girl on a team with 7 boys and the only one who knows she wants to be an engineer when she grows up. Most of them didn't want to be there at all. Their parents signed them up so they would do something besides just sports. They cheered, loudly, when they did not qualify for State because they didn't want to be at regionals at all.

**Reply by laurensvalk** [reply]
Ah, got it. From the Facebook message I figured there was some kind of deadline for the 6th, but it sounds like there is time for experimenting and learning :slightly_smiling_face:.

I hope your daughter will have a great time with her new team.

> they are between half and 3/4 of an inch thick.

What about the length? Here's a link to [that video](https://private-user-images.githubusercontent.com/12326241/345549766-7d44befb-24e3-4bbf-b602-87aebcee16cf.mp4) in case that works. I suppose turning around like this would require a longer line than that short stretch on that board screenshot.

I wouldn't necessarily try to upgrade this particular approach to a two sensor setup. This single sensor version can still be good practice in terms of detecting colors, experimenting with calibration, making turns, and so on. You don't have to build a big robot either. Just the regular educator bot with the light sensor in the middle will do.

**Comment by laurensvalk** [comment]
Following up on the other comment, I've put together a version with a single color sensor for inspiration.

We're working on a complete guide to coding and robotics, and we can certainly include examples like these.

https://github.com/pybricks/support/assets/12326241/7d44befb-24e3-4bbf-b602-87aebcee16cf

I've included the "bullet point recipe" as comment blocks throughout the program below. If you've thought out the strategy with your team, implementing each "bullet point" may be a fairly small number of blocks as shown here.

![pybricks_program_2024-07-03T17_41_09](https://github.com/pybricks/support/assets/12326241/188d4e89-36f7-4b26-a481-c3363c4ab03d)

You may also see that I've calibrated the color sensor for my "board" and sensor distance to the ground. This technique is covered [in this video](https://www.youtube.com/watch?v=kwnSbnlhOkA). If you are happy with your sensor accuracy already, you can skip that step for now.

**Reply by Schillil2001** [reply]
For some reason, the video is refusing to play, but I can at least see the code. I'll try again later when I get home from work. I'll have to modify it for 2 sensors, but one step at a time.
